### PR TITLE
feat: controllerV2 and priceProviders

### PR DESCRIPTION
**Pull request by 0xnyxos**
- Removed unused scripts from `v2-controller-data` branch
- Refactored `ControllerGenericV2`, `ChainlinkPriceProvider`, `RedstonePriceProvider`, and test suite
- Price providers consume a single price feed established at deployment time
- Price providers return bool condition and uint256 price via `conditionMet()'
- Tests written for controller and price providers

### CODE REVIEW

File: src/v2/Controllers/ControllerGeneric.sol [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+                premiumVault.finalTVL(_epochId)
+            ),
+            block.timestamp
+        );
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                                INTERNAL
+    //////////////////////////////////////////////////////////////*/
+    function _checkLiquidationConditions(
```
Comment (3xHarry) [inline src/v2/Controllers/ControllerGeneric.sol:186]: @0xnyxos looking through the internal functions, it seems that most of them have overlapping check conditions, wouldn't it make make code cleaner if we can have `MarketDoesNotExist`, `EpochNotExist` and `EpochFinishedAlready` in a seperate function or even modifier wdyt?

File: src/v2/oracles/ChainlinkPriceProvider.sol [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+} from "@chainlink/interfaces/AggregatorV3Interface.sol";
+import {
+    AggregatorV2V3Interface
+} from "@chainlink/interfaces/AggregatorV2V3Interface.sol";
+import {IVaultFactoryV2} from "../interfaces/IVaultFactoryV2.sol";
+import {IConditionProvider} from "../interfaces/IConditionProvider.sol";
+
+contract ChainlinkPriceProvider is IConditionProvider {
+    uint16 private constant _GRACE_PERIOD_TIME = 3600;
+    uint256 public constant TIME_OUT = 1 days;
```
Comment (3xHarry) [inline src/v2/oracles/ChainlinkPriceProvider.sol:15]: @0xnyxos given that the TIME_OUT window is in correlation to the heartbeat to Chainlinks heartbeat metric, == if price update is older than heartbeat we assume the price is stail and we cant use it to resolve the market, i think that we should pass this variable in the constructor when deploying. As heartbeats can vary and is defined by chainlink

**Review by 3xHarry — CHANGES_REQUESTED** [review]

**Review by 3xHarry — APPROVED** [review]

**Comment by 0xnyxos** [comment]
Additional changes in this PR:
- Deployment script for `ControllerGeneric`
- `DIAPriceProvider` contract
- `ChainlinkPriceProviderV2` contract (checking strike less than instead of more than)
- `GdaiPriceProvider` contract

File: script/v2/V2DeployContracts.s.sol [hunk 3]
```diff
@@ -21,72 +22,81 @@ contract V2DeployContracts is Script, HelperV2 {
...
         address weth = addresses.weth;
         address treasury = addresses.treasury;
         address emissionToken = addresses.y2k;
         address policy = addresses.policy;
 
-        console2.log("Address admin", addresses.admin);
-        console2.log("Address arbitrum_sequencer", addresses.arbitrum_sequencer);
-        console2.log("Address gelatoOpsV2", addresses.gelatoOpsV2);
-        console2.log("Address gelatoTaskTreasury", addresses.gelatoTaskTreasury);
+        // console2.log("Address admin", addresses.admin);
```
Comment (3xHarry) [inline script/v2/V2DeployContracts.s.sol:30]: @0xnyxos i we should prob uncomment before we merge this to main.

File: src/v2/oracles/GdaiPriceProvider.sol [hunk 4]
```diff
@@ -0,0 +1,64 @@
+/******************************************************* 
+NOTE: Development in progress by JG. Reached functional milestone; Live VST data is accessible.
```
Comment (3xHarry) [inline src/v2/oracles/GdaiPriceProvider.sol:2]: is this note still relevant?

**Review by 3xHarry — APPROVED** [review]

File: src/v2/oracles/GdaiPriceProvider.sol (hunk 4 above)
Comment (0xnyxos) [reply src/v2/oracles/GdaiPriceProvider.sol:2]: It's not - I will remove this

Have an unrelated Q - timestamps are being emitted in the genericController events. Are you happy for me to remove the timestamp as they can be implictly fetched from the block?

File: script/v2/V2DeployContracts.s.sol (hunk 3 above)
Comment (0xnyxos) [reply script/v2/V2DeployContracts.s.sol:30]: Just to clarif, we want to uncomment the console logs from admin --> gelatoTreasury?

File: src/v2/oracles/GdaiPriceProvider.sol (hunk 4 above)
Comment (3xHarry) [reply src/v2/oracles/GdaiPriceProvider.sol:2]: Indeed that makes a lot of sense, however we are the parameter in the subgraph. So for Standart V2 Controller we can leave them in, however for any other we can remvoe them, as we anyways have new ABI/ need new handlers on the subgraph.

**Pull request merged**